pretyping/coercion.ml

```diff
@@ -595,22 +588,99 @@ let inh_coerce_to_base ?loc ~program_mode env sigma j =
     sigma, res
   else (sigma, j)
 
-let inh_coerce_to_fail flags env sigma rigidonly v t c1 =
-  if rigidonly && not (Heads.is_rigid env (EConstr.Unsafe.to_constr c1) && Heads.is_rigid env (EConstr.Unsafe.to_constr t))
+let has_reverse_attribute _ = true
+
+let lookup_reversible_path_to_common_point env sigma ~src_canon ~src_coerce =
+  let src_canon, c_canon = class_of env sigma src_canon in
+  let src_coerce, c_coerce = class_of env sigma src_coerce in
+  let r1 = reachable_from c_canon in
+  let r2 = reachable_from c_coerce in
+  let r = ClTypSet.inter r1 r2 in
+  let r = ClTypSet.elements r in
+  (* XXX this order relation is not total ... *)
+  let r = List.sort (fun c1 c2 -> if ClTypSet.mem c1 (reachable_from c2) then 1 else -1) r in
```

Comment:
@pi8027 do you know what happens if List.sort is given an "non order" relation (like this one, which is not total and not anti-symmetric)?

Reply:
There is a `stable_sort` close to sort https://ocaml.org/api/List.html#VALsort so `List.sort` may or may not be stable, that's unspecified.

Reply:
In some sense, if it returns a permutation of the input given a random relation, it is still ok, if it asserts false... it is not

Reply:
That should be fine (the above doc doesn't expect a total order, any preorder is fine).

Reply:
And looking at the current implementation, it cannot throw an exception https://github.com/ocaml/ocaml/blob/trunk/stdlib/list.ml#L320

Reply:
> There is a `stable_sort` close to sort https://ocaml.org/api/List.html#VALsort so `List.sort` may or may not be stable, that's unspecified.

@proux01 Well if we start trusting implementations rather than specifications now, we see that `sort = stable_sort`... :laughing: